### ISSUE TITLE
CosineSimilarity outputs result smaller than -1

### ISSUE THREAD

**Issue by maybeLee** [keras-team/keras#17138]
**Describe the problem**.
As discussed in https://github.com/tensorflow/tensorflow/issues/58031, when I calculate the cosine similarity loss using `tf.keras.losses.CosineSimilarity` between two arrays: `[[0,2,3]]` and `[[0,2,3]]`, it outputs `-1.0000001`. Following the documentation: https://www.tensorflow.org/api_docs/python/tf/keras/losses/CosineSimilarity, the output should be within the range: [-1, 1].

**Describe the current behavior**.
The cosine similarity loss between `[[0,2,3]]` and `[[0,2,3]]` is out of [-1, 1].

**Describe the expected behavior**.
The cosine similarity loss between `[[0,2,3]]` and `[[0,2,3]]` is limited between [-1, 1].

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.
```
import tensorflow as tf
y_true = [[0.,2.,3.]]
y_pred = [[0.,2.,3.]]
cosine_loss = tf.keras.losses.CosineSimilarity()
res = cosine_loss(y_true, y_pred).numpy()
print("tensorflow: ", res)
```

**Source code / logs**.
```
tensorflow:  -1.0000001
```

**Comment by sampathweb** [comment]
It's a rounding issue with floating point numbers when calculating L2 Norm. Can you create a PR to clip the values to -1, 1 in `cosine_similarity` function -
https://github.com/keras-team/keras/blob/master/keras/losses.py#L2470
possibly by -
``` 
    y_true = tf.linalg.l2_normalize(y_true, axis=axis)
    y_pred = tf.linalg.l2_normalize(y_pred, axis=axis)
    res = -tf.reduce_sum(y_true * y_pred, axis=axis)
    return tf.clip_by_value(res, -1, 1)
```

**Comment by sampathweb** [comment]
To be consistent with other places in codebase, its better to clip via `backend.clip(res, -1, 1)` similar to how its done here - https://github.com/keras-team/keras/blob/master/keras/losses.py#L2386

**Comment by maybeLee** [comment]
Thanks for your reply, I have created a PR (https://github.com/keras-team/keras/pull/17147) following your suggestion.

**Comment by sushreebarsa** [comment]
@maybeLee Thank you for raising this issue. It will be closed once the PR is merged. Thank you!

**Comment by fchollet** [comment]
Posted on the PR:

```
An imprecision of 1e-7 is to be expected when working in float32 precision. We should not add extraneous ops such as `clip` to enforce artificial value ranges like this, because that adds computation overhead for everyone.

Here are some thing you can do:

- If you critically need to enforce a specific value range, move the `clip` op to your own application.
- Use float64 precision to reduce numerical imprecision.
- ```
